### ISSUE TITLE
Giving more detail on errors.

### ISSUE THREAD

**Issue by tiferrei** [nerdishbynature/RequestKit#38]
Hello there, 

I'm working with an API where if a code `400` is returned, it also sends JSON data containing an `error` property with a short description of the error. I was wondering if there was a way to extend the Error passed to the user to contain not only the HTTP status code but also this error message. Is this possible?

Thank you,
Tiago

**Comment by pietbrauer** [comment]
Right now there is no such option but I would be open for a PullRequest. I solved this in [VloggerKit](https://github.com/nerdishbynature/VloggerKit/blob/master/VloggerKit/Videos.swift#L4-L39) by using my own method or not using the `default override`. A more general option is definitely appreciated.

I would opt to return the error and attach the JSON in the errors `userInfo` with a custom key like Alamofire does it.

**Comment by tiferrei** [comment]
I've been trying to adapt VloggerKit's implementation to basically return a simple enum that inherits from both `String` and `Error`, so that I can provide almost like keys for each possible error. So far it;s been looking a bit like this:
```swift
/// Wrapper for the Rooms response
@objc open class RoomsResponse: NSObject, Codable {
    open var OK: Bool?
    open var error: String?
    open var rooms: [Room]?

    enum CodingKeys: String, CodingKey {
        case OK = "ok"
        case error
        case rooms
    }
}

/// Error options for the Rooms response

public enum RoomsError: String, Error {
    case NoTokenProvided = "No token provided"
    case TokenDoesNotExist = "Token does not exist"
    
    public static var errorDomain: String {
        return UCLKitErrorDomain
    }
    
    public var errorCode: Int {
        return 400
    }
}

/// Payload from the Rooms response
@objc open class Room: NSObject, Codable {
    @objc open var roomID: String?
    @objc open var roomName: String?
    @objc open var siteID: String?
    @objc open var siteName: String?
    open var capacity: Int?
    open var classification: Classification?
    open var automated: Automation?
    @objc open var location: Location?

    enum CodingKeys: String, CodingKey {
        case roomID = "roomid"
        case roomName = "roomname"
        case siteID = "siteid"
        case siteName = "sitename"
        case capacity
        case classification
        case automated
        case location
    }
}

public func rooms(_ session: RequestKitURLSession = URLSession.shared, roomID: String = "", roomName: String = "", siteID: String = "", siteName: String = "", classification: Classification = Classification.Unknown, capacity: String = "", completion: @escaping (_ response: Response<[Room]>) -> Void) {
        let router = RoomBookingsRouter.readRooms(configuration, roomID, roomName, siteID, siteName, classification, capacity)
        guard let request = router.request() else {
            let error = NSError(domain: UCLKitErrorDomain, code: 404, userInfo: nil)
            completion(Response.failure(error))
            return
        }
    
        let task = session.dataTask(with: request) { data, response, error in
            if let response = response as? HTTPURLResponse {
                if response.statusCode == 400 {
                    if let data = data, let error = error {
                        do {
                            let decoder = JSONDecoder()
                            let decoded = try decoder.decode(RoomsResponse.self, from: data)
                            if let detailedError = RoomsError(rawValue: decoded.error!) as Error? {
                                completion(Response.failure(detailedError))
                            } else {
                                completion(Response.failure(error))
                            }
                            
                        } catch {
                            completion(Response.failure(error))
                        }
                    }
                    return
                }
            }
            
            if let error = error {
                completion(Response.failure(error))
            } else {
                if let data = data {
                    do {
                        let decoder = JSONDecoder()
                        let decoded = try decoder.decode(RoomsResponse.self, from: data)
                        completion(Response.success(decoded.rooms!))
                    } catch {
                        completion(Response.failure(error))
                    }
                }
            }
        }
        task.resume()
    }
```

I can't however, figure out how to adapt RequestKit to initialise an error from a generic type. And also don't know how to provide a generic solution. The dev would have to provide some kind of key for where the error would be returned.

Any tips on this?

Cheers

**Comment by tiferrei** [comment]
I think it would be cool to let devs pass something other than the `NSError`, this way if the dev can't be bothered to write custom error handling, he can pass the `NSError`, otherwise, he could also pass a custom error type. I think now the main thing preventing this is that a failure must complete with `Error`, while a success can pass a generic. What do you think @pietbrauer?

**Comment by pietbrauer** [comment]
Yes, this could be done in the `userInfo` too. No need for a second object passed back in the failure case.

**Comment by tiferrei** [comment]
Code pushed and will be released with 2.2.0. Closing.
